Esp8266/Esp8266.ino

```diff
@@ -1,86 +1,116 @@
+    pb_decode(&input_stream, esp8266_Esp8266_Command_fields, &command);
+  }
+} input;
 
-//  WiFi.mode(WIFI_STA);
-  WiFi.begin(ssid, password);
+struct OutputData {
+  static const unsigned buffer_length = esp8266_Esp8266_Metrics_size;
+  unsigned char buffer[buffer_length];
+  pb_ostream_t output_stream;
+  esp8266_Esp8266_Metrics metrics;
 
-  while (WiFi.status() != WL_CONNECTED) {
-    Serial.println("WIFI connection failed, reconnecting...");
-    delay(500);
+  OutputData() { refresh(); }
+  void refresh() {
+    output_stream = pb_ostream_from_buffer(buffer, buffer_length);
   }
-  
-  Serial.print("\nWiFi connected, ");
-  Serial.print("IP address: ");
-  Serial.println(WiFi.localIP());
+
+  void encodeAndWrite() {
+    refresh();
+    pb_encode(&output_stream, esp8266_Esp8266_Metrics_fields, &metrics);
+    client.write(buffer, output_stream.bytes_written);
+  }
+} output;
+
+void setupWiFi();
+void printConnection(const char * name, IPAddress address, int port);
+
+void handleData();
+
+// setup WIFI and sensor
+void setup() {
+  Serial.begin(9600);
+  setupWiFi();
 }
 
 
 void loop() {
-  digitalWrite(LED_BUILTIN, LOW);
-  Serial.print("connecting to ");
-  Serial.println(addr);
-  client.connect(addr, port);
-
-  if (!client.connect(addr, port)) {
-    Serial.println("connection failed");
-    Serial.println("wait 5 sec to reconnect...");
-    delay(5000);
-    return;
+  client = server.available();
+  if (client) {
+    printConnection("controller", client.remoteIP(), client.remotePort());
+    client.setNoDelay(true);
+    while (client.connected()) {
+      if (client.available()) {
+      input.readAndDecode();
+      handleData();
+      output.encodeAndWrite();
+      }
+    }
+    Serial.println("Disconnected");
   }
-    
-  pb_Event temp = pb_Event_init_zero;
-  temp.stat = 1000;
-  
-  sendStat(temp);
-  digitalWrite(LED_BUILTIN, HIGH);
-  
-  delay(5000);
 }
 
-void sendStat(pb_Event e) {
-  uint8_t buff[128];
-  pb_ostream_t stream = pb_ostream_from_buffer(buff, sizeof(buff));
-  char* consoleStatusInfo = (char*)buff;
-  if (!pb_encode(&stream, pb_Event_fields, &e)){
-    Serial.println("failed to encode temp proto");
-    Serial.println(PB_GET_ERROR(&stream));
-    return;
-  } else {
-    Serial.println("Encoded Sucessfuly");
+void setupWiFi() {
+  Serial.print("Attempting to connect to WiFi SSID: ");
+  Serial.print(ssid);
+
+  WiFi.begin(ssid, pass);
+  while (WiFi.status() != WL_CONNECTED) {
+    delay(500);
+    Serial.print(".");
   }
-//  Serial.printf("Sending info for...%d\nMessage String : %s\n", e.stat, consoleStatusInfo);
-  Serial.print("Message Length: ");
-  Serial.println(stream.bytes_written);
- 
-  Serial.print("Message: ");
- 
-  for(int i = 0; i<stream.bytes_written; i++){
-    Serial.printf("%02X",buff[i]);
+  Serial.println();
+  server.begin();
+  printConnection("WiFi", WiFi.localIP(), port);
+}
+
+void printConnection(const char * name, IPAddress address, int port) {
```

Comment:
Very cool. I like the splitting up of tasks.

Reply:
Its mostly cuz I rewrote the same code twice, once for client and once for WiFi.